File: turn-orchestrator/src/states/tearing_down.rs
//! `tearing_down` handler: stop sandbox if any, then transition to stopped.

use harness_types::AgentEvent;
use iii_sdk::{TriggerRequest, III};
use serde_json::json;

use crate::events;
use crate::persistence;
use crate::state::{TurnState, TurnStateRecord};

pub async fn handle(iii: &III, record: &mut TurnStateRecord) -> anyhow::Result<()> {
    let request = persistence::load_run_request(iii, &record.session_id).await;
    if crate::states::assistant::approval_required_enabled(&request) {
        match crate::states::functions::consume_resolved_approval_entries(iii, &record.session_id)
            .await
        {
            Ok(prepared) if !prepared.is_empty() => {
                let executed =
                    crate::states::functions::executed_staging_for_new_prepare_batch(&[]);
                persistence::save_executed_calls(iii, &record.session_id, &executed).await;
                persistence::save_prepared_calls(iii, &record.session_id, &prepared).await;
                record.transition_to(TurnState::FunctionExecute);
                return Ok(());
            }
            Ok(_) => {}
            Err(err) => {
                tracing::warn!(
                    %err,
                    session_id = %record.session_id,
                    "approval::consume failed during teardown; stopping session"
                );
            }
        }
    }

    if let Some(sandbox_id) = persistence::load_sandbox_id(iii, &record.session_id).await {
        if let Err(e) = iii
            .trigger(TriggerRequest {
                function_id: "sandbox::stop".into(),
                payload: json!({ "sandbox_id": sandbox_id, "wait": true }),
                action: None,
                timeout_ms: Some(60_000),
            })
            .await
        {
            tracing::warn!(error = %e, sandbox_id = %sandbox_id, "sandbox::stop failed during teardown");
        }
    }

    let messages = persistence::load_messages(iii, &record.session_id).await;
    events::emit(iii, &record.session_id, &AgentEvent::AgentEnd { messages }).await;

    record.transition_to(TurnState::Stopped);
    Ok(())
}

File: harness-node/src/turn-orchestrator/states/tearing-down.ts
/**
 * `turn::tearing_down`. Stop sandbox, emit `agent_end`, transition to `stopped`.
 *
 * **Incoming**: flat `{ session_id }` via FIFO enqueue on `turn-step`.
 * **Outgoing**: `{ ok, from_state, to_state }` on success; stale skip when state drifted.
 */

import type { ISdk } from '../../runtime/iii.js';
import { logger } from '../../runtime/otel.js';
import type { AgentMessage } from '../../types/agent-message.js';
import { emit } from '../events.js';
import * as persistence from '../persistence.js';
import { type TurnStateRecord, transitionTo } from '../state.js';
import {
  TurnStepPayloadSchema,
  type TurnStepPayload,
  type TurnStepResult,
  staleSkipResult,
} from '../turn-step-payload.js';

type SandboxStopPayload = { sandbox_id: string; wait: true };

export async function handleTearingDown(iii: ISdk, rec: TurnStateRecord): Promise<void> {
  const sandbox_id = await persistence.loadSandboxId(iii, rec.session_id);
  if (sandbox_id) {
    try {
      await iii.trigger<SandboxStopPayload, unknown>({
        function_id: 'sandbox::stop',
        payload: { sandbox_id, wait: true },
        timeoutMs: 60_000,
      });
    } catch (err) {
      logger.warn('sandbox::stop failed during teardown', {
        sandbox_id,
        err: String(err),
      });
    }
  }
  const messages: AgentMessage[] = await persistence.loadMessages(iii, rec.session_id);
  await emit(iii, rec.session_id, { type: 'agent_end', messages });
  transitionTo(rec, 'stopped');
}

export async function execute(iii: ISdk, payload: TurnStepPayload): Promise<TurnStepResult> {
  const rec = await persistence.loadRecord(iii, payload.session_id);
  if (!rec) {
    throw new Error(`turn::tearing_down invariant: missing session ${payload.session_id}`);
  }
  const skipped = staleSkipResult('tearing_down', rec);
  if (skipped) return skipped;

  const from_state = rec.state;
  try {
    await handleTearingDown(iii, rec);
  } catch (err) {
    throw new Error(`transition from ${from_state} failed: ${String(err)}`);
  }
  await persistence.saveRecord(iii, rec);
  return { ok: true, from_state, to_state: rec.state };
}

export function register(iii: ISdk): void {
  iii.registerFunction(
    'turn::tearing_down',
    async (payload: unknown) => execute(iii, TurnStepPayloadSchema.parse(payload)),
    {
      description:
        'Run one durable FSM transition for session in state tearing_down: stop sandbox and mark stopped.',
    },
  );
}
